client/extensions/woocommerce/app/dashboard/index.js

```diff
@@ -50,49 +50,61 @@ class Dashboard extends Component {
 		className: PropTypes.string,
 		finishedInitialSetup: PropTypes.bool,
 		hasOrders: PropTypes.bool,
+		isSetupComplete: PropTypes.bool,
 		loading: PropTypes.bool,
 		selectedSite: PropTypes.shape( {
 			ID: PropTypes.number.isRequired,
 			slug: PropTypes.string.isRequired,
 			URL: PropTypes.string.isRequired,
 		} ),
+		siteId: PropTypes.number,
 		mailChimpConfigured: PropTypes.bool,
 		fetchOrders: PropTypes.func,
 		fetchSetupChoices: PropTypes.func,
 		requestSyncStatus: PropTypes.func,
+		setupChoicesLoading: PropTypes.bool,
 	};
 
 	componentDidMount = () => {
-		const { selectedSite, productsLoaded } = this.props;
+		const { siteId, productsLoaded } = this.props;
 
-		if ( selectedSite && selectedSite.ID ) {
-			this.props.fetchSetupChoices( selectedSite.ID );
-			this.props.fetchOrders( selectedSite.ID );
-			this.props.requestSettings( selectedSite.ID );
+		if ( siteId ) {
+			this.props.fetchSetupChoices( siteId );
+			this.props.fetchOrders( siteId );
+			this.props.requestSettings( siteId );
 
 			if ( ! productsLoaded ) {
-				this.props.fetchProducts( selectedSite.ID, { page: 1 } );
+				this.fetchSiteProducts( siteId );
 			}
 		}
 	};
 
 	componentWillReceiveProps = newProps => {
-		const { selectedSite, productsLoaded } = this.props;
+		const { siteId, productsLoaded } = this.props;
 
-		const newSiteId = newProps.selectedSite ? newProps.selectedSite.ID : null;
-		const oldSiteId = selectedSite ? selectedSite.ID : null;
+		const newSiteId = newProps.siteId ? newProps.siteId : null;
+		const oldSiteId = siteId ? siteId : null;
 
 		if ( newSiteId && oldSiteId !== newSiteId ) {
 			this.props.fetchSetupChoices( newSiteId );
 			this.props.fetchOrders( newSiteId );
 			this.props.requestSettings( newSiteId );
 
 			if ( ! productsLoaded ) {
-				this.props.fetchProducts( newSiteId, { page: 1 } );
+				this.fetchSiteProducts( newSiteId );
 			}
 		}
 	};
 
+	fetchSiteProducts = siteId => {
+		const params = { page: 1 };
+		const failureAction = dispatch => {
+			dispatch( this.props.fetchProductsFailure( siteId, params ) );
+		};
```

Comment:
This is very strange. `this.props.fetchProductsFailure` is a bound action creator. That means calling it creates the action and immediately dispatches it.

Calling `dispatch` generally returns the just-dispatched action object (if not intercepted by some middleware). Your explicit `dispatch` call therefore dispatches the failure action for a second time.

I think what you really want here is just create the failure action and pass it as parameter to `fetchProducts`:
```js
// call the imported action creator. It creates an action object.
// It's all just data, nothing gets executed.
const failureAction = fetchProductsFailure( siteId, params );
// call the bound action creator. This calls `dispatch` internally and executes the request.
this.props.fetchProducts( siteId, params, null, failureAction );
```